### PR TITLE
migrate testing from qunit to mocha

### PR DESCRIPTION
**Pull request by patrickkettner**

### CODE REVIEW

**Comment by patrickkettner** [comment]
I almost have the sauce job working - there is an issue with selenium on saucelabs on Opera 12 that I am speaking with their support about, and there are a few timeout issues with old IE. Also a few caniuse mismatches on a couple browsers.

File: Gruntfile.js [hunk 1]
```diff
@@ -45,54 +53,120 @@ module.exports = function( grunt ) {
...
+            return [
+              require('connect-jade-static')({
+                baseDir: options.base,
+                baseUrl: '/',
+                jade: {
+                  unitTests: browserTests,
+                  integrationTests: grunt.file.expand(['test/browser/integration/*.js'])
+                }
+              }),
+              function(req, rep, next) {
```
Comment (ryanseddon) [inline Gruntfile.js]: Nit picky, but my brain had to pause for a second to understand what `rep` was I think `res` is more obvious.

File: feature-detects/mathml.js [hunk 2]
```diff
@@ -21,8 +21,8 @@ define(['Modernizr', 'testStyles'], function( Modernizr, testStyles ) {
   Modernizr.addTest('mathml', function() {
     var ret;
 
-    testStyles('#modernizr{position:absolute}', function(node){
-      node.innerHTML = '<math><mfrac><mi>xx</mi><mi>yy</mi></mfrac></math>';
+    testStyles('#modernizr{position:absolute;display:inline-block}', function(node){
```
Comment (ryanseddon) [inline feature-detects/mathml.js]: Do these changes fix a bug?

Comment (patrickkettner) [inline feature-detects/mathml.js]: yesir. (I presume) the detect came out of [this conversation](https://github.com/mathjax/MathJax/issues/182). The issue here is that we are injecting into a div with block styling. so depending on the styling of the page, the resulting div will be much wider than the height. On my testing page on FF35 via a MBPr, the height was 30px, and the width was 960px.

I actually think just checking the height of the element would be sufficient, but I imagine that the fine people at MathJax know much more than I do about support.

File: Gruntfile.js (hunk 1 above)
Comment (patrickkettner) [inline Gruntfile.js]: updated

File: feature-detects/mathml.js (hunk 2 above)
Comment (patrickkettner) [inline feature-detects/mathml.js]: I noticed it when validating the caniuse suite in firefox (I just added mathml to the `map`)

File: test/browser/integration/global.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+describe('basics', function() {
+
+  it('createa a global modernizr object', function() {
```
Comment (ryanseddon) [inline test/browser/integration/global.js]: `createa a` -> `creates a`

Comment (patrickkettner) [inline test/browser/integration/global.js]: fixed

File: test/browser/integration/prefixed.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  // https://gist.github.com/523692
+  function gimmePrefix(prop, obj){
+    var prefixes = ['Moz','Khtml','Webkit','O','ms'],
+      domPrefixes = ['moz','khtml','webkit','o','ms'],
+      elem     = document.createElement('div'),
+      upper    = prop.charAt(0).toUpperCase() + prop.slice(1),
+      len;
+
+    if(!obj) {
+      if (prop in elem.style)
```
Comment (ryanseddon) [inline test/browser/integration/prefixed.js]: Should we just stick to a single if convention an require curly braces all the time?

File: test/browser/integration/prefixed.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        return prop;
+
+      for (len = prefixes.length; len--; ){
```
Comment (ryanseddon) [inline test/browser/integration/prefixed.js]: There's a few files that have space before `{` in for loops. Don't mind either.

File: test/browser/integration/prefixed.js (hunk 4 above)
Comment (patrickkettner) [inline test/browser/integration/prefixed.js]: yeah, most definitely. almost all of `integration` was copied over fairly blindly.
I'll clean up what I see, and after this merges I will work on getting some jscs chocolate all up in our peanut butter

Comment (patrickkettner) [inline test/browser/integration/prefixed.js]: updated everything I saw that was without braces

File: test/browser/src/injectElementWithStyles.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    var callback = function() {
+      var modernizr = document.getElementById('modernizr');
+      return modernizr.clientWidth === 10;
+    };
+
+    var result = injectElementWithStyles('#modernizr{width: 10px}', callback);
+    expect(result).to.be(true);
+  });
+
+  it('passses back a rule matching what we gave it', function(done) {
```
Comment (ryanseddon) [inline test/browser/src/injectElementWithStyles.js]: too many `s`'s

Comment (patrickkettner) [inline test/browser/src/injectElementWithStyles.js]: fixeddd

File: test/browser/src/isEventSupported.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    expect(isEventSupported()).to.be(false);
+  });
+
+
+  it('returns true when the event exists', function() {
+    expect(isEventSupported('click')).to.be(true);
+  });
+
+  it('returns false when the event does not exists', function() {
+    expect(isEventSupported('fart')).to.be(false);
```
Comment (ryanseddon) [inline test/browser/src/isEventSupported.js]: I laughed at this way more than a grown adult should

Comment (patrickkettner) [inline test/browser/src/isEventSupported.js]: taint the only one ;)

File: test/browser/src/load.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        expect(yepnope.calledOnce).to.be(true);
+
+        console.log = log;
+        console.warn = warn;
+      });
+
+      afterEach(function() {
+        yepnope = undefined;
+        // ie <= 8 doesn't support deleting window properties,
+        //  so wefallback to setting it to undefiend
```
Comment (ryanseddon) [inline test/browser/src/load.js]: `wefallback` -> `we fallback`

Comment (patrickkettner) [inline test/browser/src/load.js]: fixed

File: test/browser/src/prefixed.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    expect(atRule.calledOnce).to.be(true);
+  });
+
+  it('uses cssToDOM to lookup rules with "-"', function() {
+    expect(prefixed('fake-rule')).to.equal('fakeRule');
+    expect(cssToDOM.calledOnce).to.be(true);
+    expect(testPropsAll.calledOnce).to.be(true);
+  });
+
+  it('uses looks up properties on an element when provided', function() {
```
Comment (ryanseddon) [inline test/browser/src/prefixed.js]: this reads weird

Comment (patrickkettner) [inline test/browser/src/prefixed.js]: yeah, that is a typo mixed with poor grammar. How is "looks up properties
on an element, when one is provided"

File: test/browser/src/prefixes.js [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+  describe('prefixes enabled', function() {
+    before(function(done) {
+      setup(done, true);
+    });
+
+    after(teardown);
+
+    it('returns prefixes', function(done) {
+
```
Comment (ryanseddon) [inline test/browser/src/prefixes.js]: remove newline

File: test/browser/src/setClasses.js [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    it('adds a class with a prefix', function(done) {
+      requirejs(['setClasses'], function(setClasses) {
+        setClasses(['detect']);
+        expect(elm.className).to.contain('fakedetect');
+        done();
+      });
+    });
+
+    after(teardown);
+
```
Comment (ryanseddon) [inline test/browser/src/setClasses.js]: remove new line

File: test/node/lib/build.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+    it('should exclude options that are not requested', function(done) {
+      build({'feature-detects': ['dom/classlist']}, function(file) {
+        expect(file).to.contain('classlist');
+        expect(file).to.not.contain('boxsizing');
+        done();
+      });
+    });
+
+    it('should stip out DOC comments when `uglify`ing', function(done) {
```
Comment (ryanseddon) [inline test/node/lib/build.js]: `stip` -> `strip`

File: feature-detects/mathml.js (hunk 2 above)
Comment (ryanseddon) [inline feature-detects/mathml.js]: yay for tests finding issues, who would of thought.

File: test/browser/src/setClasses.js (hunk 11 above)
Comment (patrickkettner) [inline test/browser/src/setClasses.js]: sorry - which newline?

File: test/browser/src/isEventSupported.js (hunk 7 above)
Comment (ryanseddon) [inline test/browser/src/isEventSupported.js]: Yeah i notived the consistency of fart throught.

File: test/browser/src/prefixed.js (hunk 9 above)
Comment (ryanseddon) [inline test/browser/src/prefixed.js]: perfect

**Comment by ryanseddon** [comment]
This is an amazing effort!

File: test/browser/src/isEventSupported.js (hunk 7 above)
Comment (patrickkettner) [inline test/browser/src/isEventSupported.js]: trying to keep it profesh

File: test/browser/src/setClasses.js (hunk 11 above)
Comment (ryanseddon) [inline test/browser/src/setClasses.js]: ah oops I thought that was the end of the describe. nvm.

File: test/browser/src/setClasses.js [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    before(function(done) {
+      setup(done, {
+        'enableClasses': true
+      });
+    });
+
+    after(teardown);
+
+    it('adds a class without a prefix', function(done) {
+
```
Comment (ryanseddon) [inline test/browser/src/setClasses.js]: Actually I meant this one

**Comment by ryanseddon** [comment]
We should still build a static test page so we can link people to the suite. Should just generate the static files in the travis build

**Comment by patrickkettner** [comment]
yep - I realized that after I pushed it up yesterday. What is here is 98% static. the jade files should be compiled to html, and the ua-parser can't be served dynamically like it is currently. otherwise, we would be golden. 

the ua parsing thing may be a little more tricky. short term I think I will just revert to paul's browserified parser currently used. long term see if we can get uaparser browsified proper, so that we can include as a node module and update throughout time.

**Comment by ryanseddon** [comment]
yep sounds good

**Comment by patrickkettner** [comment]
after the static changes are made, are you :+1:?

**Comment by ryanseddon** [comment]
yep :+1:

**Comment by patrickkettner** [comment]
@ryanseddon wanna give ita quick once over? Basically, on successful merge to master, the `gh-pages` branch is updated with the output of `grunt copy:gh-pages`

huge thanks to @alrra for showing me that and getting it working

**Comment by ryanseddon** [comment]
Yep looks good, so `grunt copy:gh-pages` will just be ran manually after any changes to master?

Also appveyor failed due to phatomjs timing out is this a blocker for merging?

**Comment by ryanseddon** [comment]
Also seeing 4 tests failing on Fx35

**Comment by patrickkettner** [comment]
No, it runs automatically on Travis. We don't have to manually do anything
ever again

**Comment by patrickkettner** [comment]
@ryanseddon what tests are you seeing failed on what system? I am all green on FF35 on Yosemite

**Comment by patrickkettner** [comment]
@ryanseddon able to reproduce, weird caching thing I guess. One issue is related to this, three are not. I was calling `!!bool` rather than `bool.valueOf()`, which meant that `new Boolean(false)` objects where coming back as true. Still debugging appveyor.

**Comment by SlexAxton** [comment]
Magic. You two.

**Pull request merged**